### PR TITLE
feat(eslint-plugin): add `no-implicit-any-catch` rule

### PR DESCRIPTION
**Pull request by phiresky**
In the next TS version, adding both `catch (e: unknown) {}` and `catch (e: any) {}` will be supported.

This PR adds support for enforcing specifying a type for catch clauses. This is basically somethinng that should be caught by the `noImplicitAny` compiler flag, but probably won't be added to that because of backwards compatibility.

See https://github.com/microsoft/TypeScript/issues/36775 which was fixed here: https://github.com/microsoft/TypeScript/pull/39015

This PR also adds support for the new syntax to  typescript-estree.

The following pattern is considered a warning:

```ts
try {
  // ...
} catch (e) { // warning: Implicit any in catch clause
  // ...
}
```

The following patterns are not warnings:

```ts
try {
  // ...
} catch (e: unknown) {
  // ...
}
```

```ts
try {
  // ...
} catch (e: any) {
  // ...
}
```

This PR will throw some "This snippet should be formatted correctly. Use the fixer to format the code" errors since this syntax is not yet supported in the formatter the lint is using.

To make this work, you need to set the TypeScript dependency to nightly (one containing this fix, wait until tomorrow :))

### CODE REVIEW

**Comment by phiresky** [comment]
Also, since prettier does not support this syntax yet, the autoformatter currently strips the types from the docs etc on auto-format, which is pretty unhandy.

File: packages/typescript-estree/src/convert.ts [hunk 1]
```diff
@@ -697,7 +697,7 @@ export class Converter {
         return this.createNode<TSESTree.CatchClause>(node, {
           type: AST_NODE_TYPES.CatchClause,
           param: node.variableDeclaration
-            ? this.convertChild(node.variableDeclaration.name)
+            ? this.convertChild(node.variableDeclaration).id
```
Comment (bradzacher) [inline packages/typescript-estree/src/convert.ts:700]: now that `catch (x: any)` is valid syntax, could you also add some tests for this?
Just add a few into this folder so we can validate that the type annotation is generated correctly.

https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/shared-fixtures/fixtures/typescript/basics

Comment (phiresky) [reply packages/typescript-estree/src/convert.ts:700]: Done. It now throws 

```
@typescript-eslint/typescript-estree: FAIL tests/ast-alignment/spec.ts
@typescript-eslint/typescript-estree:   ● TEST FAIL [BABEL ERRORED, BUT TS-ESTREE DID NOT] - /home/tehdog/data/dev/2020/typescript-eslint/packages/shared-fixtures/fixtures/typescript/basics/catch-type-any.src.ts
@typescript-eslint/typescript-estree:     expect(received).toEqual(expected) // deep equality
@typescript-eslint/typescript-estree:     Expected: [SyntaxError: Unexpected token, expected ")" (3:10)
@typescript-eslint/typescript-estree:       1 | try {
@typescript-eslint/typescript-estree:       2 |·
@typescript-eslint/typescript-estree:     > 3 | } catch (e: any) {
@typescript-eslint/typescript-estree:         |           ^
@typescript-eslint/typescript-estree:       4 |·
@typescript-eslint/typescript-estree:       5 | }
@typescript-eslint/typescript-estree:       6 | ]
@typescript-eslint/typescript-estree:     Received: null
@typescript-eslint/typescript-estree:       33 |     if (!typeScriptESTreeResult.parseError) {
@typescript-eslint/typescript-estree:       34 |       it(`TEST FAIL [BABEL ERRORED, BUT TS-ESTREE DID NOT] - ${filename}`, () => {
@typescript-eslint/typescript-estree:     > 35 |         expect(typeScriptESTreeResult.parseError).toEqual(
@typescript-eslint/typescript-estree:          |                                                   ^
@typescript-eslint/typescript-estree:       36 |           babelParserResult.parseError,
@typescript-eslint/typescript-estree:       37 |         );
@typescript-eslint/typescript-estree:       38 |       });
@typescript-eslint/typescript-estree:       at Object.<anonymous> (tests/ast-alignment/spec.ts:35:51)
@typescript-eslint/typescript-estree:   ● TEST FAIL [BABEL ERRORED, BUT TS-ESTREE DID NOT] - /home/tehdog/data/dev/2020/typescript-eslint/packages/shared-fixtures/fixtures/typescript/basics/catch-type-unknown.src.ts
@typescript-eslint/typescript-estree:     expect(received).toEqual(expected) // deep equality
@typescript-eslint/typescript-estree:     Expected: [SyntaxError: Unexpected token, expected ")" (3:10)
@typescript-eslint/typescript-estree:       1 | try {
@typescript-eslint/typescript-estree:       2 |·
@typescript-eslint/typescript-estree:     > 3 | } catch (e: unknown) {
@typescript-eslint/typescript-estree:         |           ^
@typescript-eslint/typescript-estree:       4 |·
@typescript-eslint/typescript-estree:       5 | }
@typescript-eslint/typescript-estree:       6 | ]
@typescript-eslint/typescript-estree:     Received: null
@typescript-eslint/typescript-estree:       33 |     if (!typeScriptESTreeResult.parseError) {
@typescript-eslint/typescript-estree:       34 |       it(`TEST FAIL [BABEL ERRORED, BUT TS-ESTREE DID NOT] - ${filename}`, () => {
@typescript-eslint/typescript-estree:     > 35 |         expect(typeScriptESTreeResult.parseError).toEqual(
@typescript-eslint/typescript-estree:          |                                                   ^
@typescript-eslint/typescript-estree:       36 |           babelParserResult.parseError,
@typescript-eslint/typescript-estree:       37 |         );
@typescript-eslint/typescript-estree:       38 |       });
@typescript-eslint/typescript-estree:       at Object.<anonymous> (tests/ast-alignment/spec.ts:35:51)
```

which i guess is good?

Comment (bradzacher) [reply packages/typescript-estree/src/convert.ts:700]: yup that's to be expected; babel doesn't support it, and babel throws for things it doesn't support

You can ignore the fixture you added from the babel AST alignment tests here:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/tests/ast-alignment/fixtures-to-test.ts

File: packages/eslint-plugin/docs/rules/no-implicit-any-catch.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Disallow usage of the implicit `any` type in catch clauses (`no-implicit-any-catch`)
+
+Using the `any` type defeats the purpose of using TypeScript.
+When `any` is used, all compiler type checks around that value are ignored.
+
+The noImplicitAny flag in TypeScript does not cover this due to backwards compatibility reasons.
+
+## Rule Details
+
+This rule doesn't requires an explicit type to be declared in the catch clause error argument.
```
Comment (glen-84) [inline packages/eslint-plugin/docs/rules/no-implicit-any-catch.md:10]:
```suggestion
This rule requires an explicit type to be declared in the catch clause error argument.
```

File: packages/typescript-estree/src/convert.ts (hunk 1 above)
Comment (phiresky) [reply packages/typescript-estree/src/convert.ts:700]: i've added them to the ignored fixtures

**Comment by tom-sherman** [comment]
Can we have an option on this that enforces `unknown` as the argument type in this PR? Or should that be discussed in a seperate issue?

**Comment by bradzacher** [comment]
> Can we have an option on this that enforces `unknown` as the argument type in this PR? Or should that be discussed in a seperate issue?

https://github.com/typescript-eslint/typescript-eslint/pull/2202/files#diff-edf249e9f9f8af86a2cf0ed9bb7c9a3dR42-R48

**Comment by phiresky** [comment]
It does cover that, but I've defaulted it to false since I think you'd usually want to use the `no-explicit-any` rule for that since it covers all of those cases.

**Comment by tom-sherman** [comment]
@bradzacher 

> doesn't this PR already cover that?

You're right, I missed that part - nice!

File: packages/eslint-plugin/docs/rules/no-implicit-any-catch.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  // ...
+}
+```
+
+The following patterns are not warnings:
+
+```ts
+try {
+  // ...
+} catch (e) {
```
Comment (WORMSS) [inline packages/eslint-plugin/docs/rules/no-implicit-any-catch.md:27]:
```suggestion
} catch (e: any) {
```

File: packages/eslint-plugin/docs/rules/no-implicit-any-catch.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  // ...
+}
+```
+
+```ts
+try {
+  // ...
+} catch (e) {
```
Comment (WORMSS) [inline packages/eslint-plugin/docs/rules/no-implicit-any-catch.md:35]:
```suggestion
} catch (e: unknown) {
```

Comment (phiresky) [reply packages/eslint-plugin/docs/rules/no-implicit-any-catch.md:35]: this is sadly impossible to commit as long as prettier styling is enforced (which automatically strips these types)

Comment (WORMSS) [reply packages/eslint-plugin/docs/rules/no-implicit-any-catch.md:35]: /* : unknown */ until then?

Comment (phiresky) [reply packages/eslint-plugin/docs/rules/no-implicit-any-catch.md:35]: good idea.

**Comment by bradzacher** [comment]
I made a few updates to fix the CI, but for some reason github isn't picking up the changes.
I committed to your master branch which this PR is tracking - so the changes are there, but unsure why GH isn't adding them to the PR.

Feel free to take a look and merge master into your branch as well.

**Comment by WORMSS** [comment]
> I committed to your master branch which this PR is tracking

You were able to push to a different users master branch?

**Comment by phiresky** [comment]
> I committed to your master branch which this PR is tracking - so the changes are there,

Wait, what? I did not know that was possible and I didn't add change any access permissions. Maybe this is a git bug since github itself didn't even notice the repo change happened?

Regardless, I've merged upstream master into my branch again so now github has picked up the changes again.

**Comment by phiresky** [comment]
Never mind, I checked the `Allow edits by maintainers` checkbox when I created the PR. And github picking it up is probably just because it was down (again) earlier today.

**Comment by bradzacher** [comment]
> You were able to push to a different users master branch?

@WORMSS  As phiresky mentioned - GH has a setting on PRs which allows maintainers of the project you're PRing to to commit to the remote branch you want to merge.

image

This permission is what enables the "update from master" button to work on the UI, but it also lets a maintainer do things like quick edit files in the UI. A keen observer also might note that this ofc means you can checkout the branch locally, make whatever changes, and push the changes.

I usually avoid doing anything beyond updating from master because it:
- removes agency and ownership from the contributor
- is a pain as their local branch will be out of sync with their remote

I usually prefer to rely upon the contributor owning e2e and making the changes that I request, but sometimes I use it in the following cases:
- If I have a minor one-line change (i.e. a documentation change), that's not worth RCing and bothering the contributor with.
- If I want to get something in for a release, I sometimes just make changes myself because I don't know when , and comment with what I did and why.

> probably just because it was down (again) earlier today

Yeah, it looks like it was just a problem with github. Seems to be working now.

**Comment by bradzacher** [comment]
okay as for the failing tests here - I think we should probably look to split out the 4.0.0 upgrade as a separate PR now.

It looks like 4.0.0 has changed how `null` types are represented in addition to the other changes.

Before, the TS parser just output `NullKeyword`, but now it outputs `LiteralType > NullKeyword`, which is a breaking change that we need to "undo".

I'll submit a PR this week to bump the version and fix all of the issues, and then we can rebase this PR to add the rule + support for type annotations on catch clause variables.

**Comment by cyrilgandon** [comment]
What about thrown Promises, should it be included in this rule too?

```
async function foo() {
}

foo().catch((err) => console.error(err));
            ---- Disallow usage of the implicit `any` type in catch clauses
```

**Comment by bradzacher** [comment]
No - that `catch` is a function call - so it is handled by things like the `no-unsafe-*` ruleset.
This rule is specifically about enforcing usage of the new TS4.0 synax.

**Pull request merged**